pkg/controller/grouping/group_entity_index.go

```diff
@@ -0,0 +1,728 @@
+// Copyright 2021 Antrea Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package grouping
+
+import (
+	"fmt"
+	"reflect"
+	"strings"
+	"sync"
+
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/klog"
+
+	"github.com/vmware-tanzu/antrea/pkg/apis/core/v1alpha2"
+	"github.com/vmware-tanzu/antrea/pkg/controller/types"
+	utilsets "github.com/vmware-tanzu/antrea/pkg/util/sets"
+)
+
+// Interface provides methods to query entities that a given group selects and groups that select a given entity.
+// It maintains indexes between groups and entities to make the query efficient. It supports callers to register
+// callbacks that will be called when a specific type of groups' entities are updated.
+type Interface interface {
+	// AddGroup adds or updates a group to the index. The caller can then get entities selected by this group.
+	AddGroup(groupType string, name string, selector *types.GroupSelector)
+	// DeleteGroup deletes a group from the index.
+	DeleteGroup(groupType string, name string)
+	// AddEventHandler registers an eventHandler for the given type of groups. When any Pod/ExternelEntity/Namespace
+	// update affects the given kind of groups, the eventHandler will be called with the affected groups.
+	AddEventHandler(groupType string, handler eventHandler)
+	// GetEntities returns the selected Pods or ExternalEntities for the given group.
+	GetEntities(groupType string, name string) ([]*v1.Pod, []*v1alpha2.ExternalEntity)
+	// GetGroupsForPod returns the groups that select the given Pod.
+	GetGroupsForPod(namespace, name string) (map[string][]string, bool)
```

Comment:
do you think we could define a new type for the group type: `type GroupType string`? It would help clarify what the returned values are if we have `map[GroupType][]string` instead of `map[string][]string`.

Reply:
Maybe an enum?

Reply:
I feel enum (int) also works, but I guess your concern is harder to track the int values when they are not in the same package?

Reply:
> Maybe I should define GroupType (using string) here and let consumers define their own values in their own packages, does it make sense to you?

That works for me as a middle ground between using a plain `string` and defining named constants for all group types in this package.

Reply:
@jianjuns  Yes, if using enum(int), they we should define them together in this package to avoid conflict.

Reply:
I think it is just like string is easier to remember and has less chance to conflict (but can still). I do not like long string lookup, but I understand your point and no strong opinion.